### PR TITLE
Fix for potential spurious wake-ups in WorkflowThreadContext

### PR DESCRIPTION
**Pull request by Spikhalskiy**
Some awaits in WorkflowThreadContext were not shielded against spurious wake-ups with condition checks or were shielded with condition checks that were not safe.

### CODE REVIEW

**Review by bergundy — COMMENTED** [review]
Are there any concerns of breaking history compatibility with this change?

**Comment by Spikhalskiy** [comment]
@bergundy nope, it's to eliminate potential rare spurious wake-ups and race conditions. Normal executions will not be affected by this change.

File: temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowThreadContext.java [hunk 1]
```diff
@@ -123,33 +130,35 @@ private void maybeEvaluateLocked(String reason) {
       } catch (Exception e) {
         evaluationFunction.apply(Throwables.getStackTraceAsString(e));
       } finally {
-        status = Status.YIELDED;
+        status = Status.RUNNING;
```
Comment (Spikhalskiy) [inline temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowThreadContext.java:133]: It's a more correct state to switch after evaluation. If running conditions are not met, we will go into YIELDED anyway after the check.

Before this change the following chains of states were possible: 
EVALUATING -> YIELDED
EVALUATING -> YIELDED (shortly) -> RUNNING and YIELDED will be set and observable without actually yielding. Ideally, we want to switch to and publish YIELDED state only when we ACTUALLY yield.

After the change, we can get 
EVALUATING -> RUNNING (shortly) -> YIELDED 
or EVALUATING -> RUNNING
and we go into YIELDED ONLY if we actually yield. This is a much more consistent state model.

**Comment by Spikhalskiy** [comment]
After thinking more, this is reduced to a small change that is not worth a separate pr, will just include it into another PR touching this class for simplicity.

**Pull request closed without merging**